### PR TITLE
Add Full-TCP test suite

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
Description
-----------
This PR adds the [Full-TCP test suite](https://github.com/aws/amazon-freertos/blob/master/libraries/abstractions/secure_sockets/test/iot_test_tcp.c#L1080) present in [amazon-FreeRTOS](https://github.com/aws/amazon-freertos) repository after some modifications to the FreeRTOS/FreeRTOS repository. This is a part of a range of additions to the FreeRTOS/FreeRTOS repository to enhance the testing capabilities of this repository. The specific modifications are listed below:
- The Full-TCP test suite uses the Secure-Socket abstraction whereas the tests being added just use FreeRTOS_Sockets since Secure Sockets is not a part of FreeRTOS/FreeRTOS.
- The test names have been changed to reflect the above changes. The names follow this convention: TCP_Sockets_\<*name of function*>_\<*TestCase*>.
- The test case execution order has been changed to make sure that a functionality (e.g. 'FreeRTOS_send') is tested before it is used in some other test-case.

**Note For reviewers**: Important changes are restricted to only 2 files, namely: `FreeRTOS_Sockets.c` and `test_tcp.c`. (Other files are just additions to the project)
OR
You can try running the demo on your local setup.

Test Steps
-----------
- Clone the Repository with `--recurse-submodules`.
- Go to 'FreeRTOS-Plus/Test/FreeRTOS_Plus_TCP_Integration_Tests/Full-TCP_tests' directory.
- NOTE: You'll need an active ethernet connection and probably need to switch off any VPN service.
- Open the `Full-TCP-tests` project.
  - Change the interface to use by modifying the 'configNETWORK_INTERFACE_TO_USE' define.
  - Change the `tcptestECHO_SERVER_ADDR[0-3]` and `tcptestECHO_PORT` in test_tcp.h to reflect the address of your echo server.
  - Save the changes.
- Compile and run the project. It should run 23 tests which should all pass.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/FreeRTOS_Sockets.c [hunk 1]
```diff
@@ -1397,6 +1397,12 @@ FreeRTOS_Socket_t *pxSocket;
 	( void ) uxOptionLength;
 
 	configASSERT( xSocket != NULL );
+
```
Comment (gkwicker) [inline FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/FreeRTOS_Sockets.c:1400]: The title of this PR is "Add Full-TCP test suite", and the description does not include the fact that it modifies FreeRTOS+TCP behavior.  Changes to FreeRTOS_Sockets.c (and in fact any FreeRTOS+TCP core files) should be in a separate PR, since these affect more than just testing.

**Comment by AniruddhaKanhere** [comment]
Closing this PR to incrementally add changes.
Will Split the PR into multiple smaller PRs.
